File: backend/app/core/upgrades/stat_modifiers/damage_1.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class Damage1(UpgradeImplementation):
    id = 'damage_1'
    name = 'Devastation'
    description = '+0.2% damage.'
    rarity = 'common'
    type = 'stat_modifier'
    cost = 2

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        appliers.modifiers.add_multiplier('attack', 'damage', 0.002)

    def validate(self, recorded_stats: dict) -> bool:
        min_expected = 10 * 1.002
        damage_per_shot = recorded_stats.get('damage_per_shot', 0)
        return damage_per_shot >= min_expected * 0.95

File: frontend/src/game/upgrades/stat_modifiers/damage_1.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const Damage1Def: UpgradeDef = {
  id: "damage_1",
  name: "Devastation",
  description: "+0.2% damage.",
  rarity: RarityID.Common,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 2,
  targetClass: UpgradeTargetID.Attack,
  fieldInTargetClass: UpgradeStatID.Damage,
  value: 0.002,
  isMultiplier: true,
  stackable: true,
  maxStacks: 99999,
}

export class Damage1 extends Upgrade {}
